### ISSUE TITLE
pretrained ResNet&VGG npy files & python3 support

### ISSUE THREAD

**Issue by JuheonYi** [Timforce/tiny-tf#6]
Hi. thanks for sharing the code!

Can I ask where/how you got the Resnet101.npy files?
I want to train other models as well, and I wonder if you could share other Resnet50, Resnet150, VGG16.npy, or some codes to generate such files.

Also by the way, I have modified your code a bit so that the code can run in python3
(seems there are some changed modules, so I fixed some)

Would there be any way to share the code as well, without merging it to your original codes?
(Obviously having code that both works for python2 and 3 would be more helpful!)

thanks

**Comment by Timforce** [comment]
Hi,
Thanks for the appreciation :)
I use [Caffe to TensorFlow][caffeTFurl] to convert the existing caffe model(with some code modification) to npy files. It's quite a useful tool.

Also, you are more than welcome to share your python3 version code! 👍
Just fork this repository and create a new branch. When it's done, please let me know and i shall add a link to your repository. 

[caffeTFurl]:https://github.com/ethereon/caffe-tensorflow

**Comment by JuheonYi** [comment]
Hi, I have forked your project, and pushed my changes as you recommended.
https://github.com/JuheonYi/tiny-tf
I modified the changes in module names, syntaxes, and so on so that the code runs in Python3.
All the codes for train/eval/demo seem to work well at the moment (I am currently training the model from scratch again to validate), and I will keep updating!

Thanks for sharing to converting tool! I will try them as well.

If you don't mind, may I ask about some other points as well?
My goal right now is to train various networks (ResNet101, ResNet50, VGG16, and possibly other ones) to observe the tradeoff in speed and accuracy.
(ResNet101 model's latency seems very long, especially for x2 upscaled input image).

However, actually this is the first time I am training the face detector myself (have only used the pre-trained ones provided in the internet), and I am having some trouble understanding some points:(

I would really appreciate it if you could give me some advice on the following questions.

1) I noticed in the codes (e.g., /lib/networks/Resnet101_train.py) that for the outputs (denoted as "FCN-fashion architecture"), two layers are obtained (score_3 & score_4), and added on after resizing score_4 for the final output.
Can you explain why the output layer is designed as this? Is this some convention in face detection networks?
For other custom models, what would be the criterion for choosing the layers to extract score_3 & score_4?

2) It seems that the output shape is [n,h,w,125], and [:,:,:,:25] becomes bbox and [:,:,:,25:] becomes the scores, after some postprocessing. Is this also the convention in face detection as well? I wonder if there are any meanings in the numbers 125 and 25.

3) Do you have any experience on training the network without any pre-training on ImageNet? I wonder if the model can be trained from scratch, without starting from pre-trained models.
(I plan to test this as well, but would be great in case you could share your experience to save time!)

Thanks again for sharing your codes! 
It really helps me a lot.

**Comment by Timforce** [comment]
Hi, 
I have add a link to your repository in README.md, thank you for sharing your code as well :)

As for your questions:
1. This is a technique known as [feature pyramid][fpyramid], which is helpful for object detection task. 
The original paper also did some [exploration][tinypaper] on different layer's corresponding actual face sizes. I believe the author choose these two layers based on the observed result and some experiments.

2. Number 25 comes from the original papar's [repo][tiny-repo], which cluster the WIDER face dataset into 25 reference bounding box sizes based on K-medroid. These sizes will later be used to generate 25 channels for confidence score and 25*4 channels for bounding box regression score. 
You can specify different number of medoids in [script][tinyurl] to generate another set of bounding box reference sizes, here i use the same set as original paper just for paper reimplementation purpose.

3. I did have trained the network from scratch before, i would say the difference is minor(if you train for long enough) but it's still there. I recommend using pre-trained weight whenever you can, this gives your network a good starting point to fine-tune itself.

Hope these idea may help you!

[fpyramid]: https://arxiv.org/abs/1612.03144
[tiny-repo]: https://github.com/peiyunh/tiny
[tinypaper]: https://arxiv.org/pdf/1612.04402.pdf
[tinyurl]: https://github.com/peiyunh/tiny/blob/master/utils/cluster_rects.m

**Comment by JuheonYi** [comment]
Hi, thank you for the detailed answers! 
Hearing your explanation, now I get what was going on inside.
I will try some other things, and update you in case I add additional functionalities in the code that may be useful.

By the way, I have tested my revised codes in Python3, and checked that the model trains fine!
(Accuracy on WIDERFACE 'hard' validation set is 0.78, very close to 0.814 of original paper).
I have updated README accordingly so that more people could have more confidence in using your codes.

Thanks a lot for the codes again! 
It's really great having a training code that I can use to train networks I want for face detection.
Really saved a lot of my time & effort.

Thanks,
Juheon
